host-ext/nacl_src/keycode_converter.h

```diff
@@ -0,0 +1,215 @@
+/* Copyright (c) 2015 The crouton Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ *
+ * Translates NaCl pp::KeyboardInputEvent::KeyCode() strings to X11 keycodes.
+ */
+
+#include <map>
+#include <string>
+#include <stdint.h>
+
+/* Values in the keycode hashmap. */
+class KeyCode {
+public:
+    KeyCode(uint8_t base, uint8_t search):
+        base_(base), search_(search) {}
+
+    KeyCode(uint8_t base): KeyCode(base, base) {}
+
+    uint8_t GetCode(const bool search_on) const {
+        if (search_on)
+            return search_;
+        else
+            return base_;
+    }
+
+private:
+    const uint8_t base_;  /* Basic keycode */
+    /* Reverse translation of keycode when Search is pressed: e.g.
+     * Search+Left => Home. In this case:
+     * base_ = Home keycode (0x6e), search_ = Left keycode (0x71) */
+    const uint8_t search_;
+};
+
+/* Class with static members only: converts KeyCode string to X11 keycode */
+class KeyCodeConverter {
+public:
+    static uint8_t GetCode(const std::string& str, const bool search_on) {
+        if (!strcodemap_)
+            InitMap();
+
+        auto it = strcodemap_->find(str);
+        /* Not found */
+        if (it == strcodemap_->end())
+            return 0;
+
+        return it->second.GetCode(search_on);
+    }
+
+private:
+    static void InitMap();
+    /* static pointer member, to avoid static variable of class type. */
+    static std::map<std::string, const KeyCode>* strcodemap_;
+};
+
+std::map<std::string, const KeyCode>* KeyCodeConverter::strcodemap_ = NULL;
```

Comment:
I know...I've always wanted to find a use for gperf though :)
I'm not even sure unordered_map is necessary.